04.ls/ls.rb

```diff
@@ -1,57 +1,149 @@
 # frozen_string_literal: true
 
 require 'optparse'
+require 'etc'
 
 COLUMN = 3
-SPACE = 2
+BLOCK_SIZE = 4
+
+FILE_TYPE = {
+  '01' => 'p',
+  '02' => 'c',
+  '04' => 'd',
+  '10' => '-',
+  '12' => 'l',
+  '14' => 's'
+}.freeze
+
+PERMISSION = {
+  '0' => '---',
+  '1' => '--x',
+  '2' => '-w-',
+  '3' => '-wx',
+  '4' => 'r--',
+  '5' => 'r-x',
+  '6' => 'rw-',
+  '7' => 'rwx'
+}.freeze
+
+IS_AUTHORITY = {
+  '0' => false,
+  '1' => true,
+  '2' => true,
+  '4' => true
+}.freeze
+
+INDEX_AUTHORITY = {
+  '1' => 8,
+  '2' => 5,
+  '4' => 2
+}.freeze
+
+AUTHORITY_TYPE = {
+  '1' => { 'x' => 't', '-' => 'T' },
+  '2' => { 'x' => 's', '-' => 'S' },
+  '4' => { 'x' => 's', '-' => 'S' }
+}.freeze
 
 def main
-  parsed_options = parse_options
-  target_dir = ARGV.empty? ? '.' : ARGV[0]
-  entry_names = Dir.entries(target_dir)
-  sorted_names = sort_names(entry_names, parsed_options)
-  filtered_names = filter_names(sorted_names, parsed_options)
-  puts output(filtered_names)
+  options = parse_options
+  target_dir_path = ARGV.empty? ? '.' : ARGV[0]
+  names = Dir.entries(target_dir_path)
+  sorted_names = sort_names(names, options)
+  filtered_names = filter_names(sorted_names, options)
+  loaded_attributes = load_attributes(filtered_names, target_dir_path) if options[:l]
+  options[:l] ? puts(format_attributes(loaded_attributes)) : puts(format_names(filtered_names))
```

Comment:
```suggestion
  options[:l] ? puts(format_attributes(filtered_names, target_dir_path)) : puts(format_names(filtered_names))
```

`options[:l]`でしか使わないデータはmainで読み込まなくていいと思います。

あと、参考演算子とputsが組み合わさると読みづらいので、いったん変数に入れる方がいいと思います。

```ruby
hoge = piyo ? ... : ...
puts hoge
```

Reply:
承知しました。
修正します！